core/src/main/scala/com/yahoo/maha/core/query/OffHeapRowList.scala

```diff
@@ -0,0 +1,194 @@
+package com.yahoo.maha.core.query
+
+import java.io.File
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicLong
+
+import com.yahoo.maha.core.query.QueryRowList.ROW_COUNT_ALIAS
+import com.yahoo.maha.core.request.Parameter.RequestId
+import com.yahoo.maha.rocksdb.{RocksDBAccessor, RocksDBAccessorBuilder}
+import com.yahoo.maha.serde.LongSerDe
+import grizzled.slf4j.Logging
+import org.apache.commons.io.{FileUtils, FilenameUtils}
+
+import scala.collection.mutable.ArrayBuffer
+import scala.util.Try
+
+/*
+    Created by pranavbhole on 2/3/20
+*/
+case class OffHeapRowListConfig(tempStoragePathOption:Option[String], inMemRowCountThreshold: Int = 10000)
+
+case class OffHeapRowList(query: Query, config: OffHeapRowListConfig) extends QueryRowList with AutoCloseable with Logging {
+
+  private[this] val tempStoragePath = OffHeapRowList.getAndValidateTempStorage(config.tempStoragePathOption)
+
+  private[this] var rocksDBAccessorOption: Option[RocksDBAccessor[Long, Row]] = None
+
+  private[this] val atomicLongRowCount = new AtomicLong(-1)
+
+  private[this] val rocksDBRowCount = new AtomicLong(0) // To be used for verification and monitoring
+
+  protected[this] val list: collection.mutable.ArrayBuffer[Row] = {
+    if(query.queryContext.requestModel.maxRows > 0) {
+      new ArrayBuffer[Row](query.queryContext.requestModel.maxRows + 10)
+    } else {
+      collection.mutable.ArrayBuffer.empty[Row]
+    }
+  }
+
+  def addRow(r: Row, er: Option[Row] = None) : Unit = {
+    postResultRowOperation(r, er)
+    if (atomicLongRowCount.get() >= config.inMemRowCountThreshold - 1) {
+      persist(r)
+    } else {
+      list+=r
+      atomicLongRowCount.incrementAndGet()
+    }
+  }
+
+  private[this] def persist(row:Row): Unit =  {
+    if(rocksDBAccessorOption.isEmpty) {
+      rocksDBAccessorOption = Some(OffHeapRowList.initRocksDB(tempStoragePath, query))
+    }
+    rocksDBRowCount.incrementAndGet()
+    rocksDBAccessorOption.get.put(atomicLongRowCount.incrementAndGet(), row)
+  }
+
+  private[this] def rocksDBGet(index:Long): Row = {
+    require(index <= atomicLongRowCount.get(), s"Failed to get the index ${index} from rocksdb")
+    require(rocksDBAccessorOption.isDefined, "Invalid get call, RocksDB is yet not initialized")
+    val rowOption = rocksDBAccessorOption.get.get(index)
+    require(rowOption.isDefined, s"Failed to get row at index ${index} from rocksDB")
+    rowOption.get
+  }
+
+  def isEmpty : Boolean = {
+    if (atomicLongRowCount.get() < 0) {
+      true
+    } else false
+  }
+
+  override protected def kill(): Unit = {
+    info(s"called killed on OffHeapRowList, Stats:-> inMemRowCountThreshold: ${config.inMemRowCountThreshold}, sizeOfInMemStore: ${sizeOfInMemStore()}, sizeOfRocksDB:${sizeOfRocksDB()}")
```

Comment:
Probably want this in debug mode right, not on every query ?

Reply:
printing stats only at debug or rocksdb init case.